File: flutter_app/lib/features/suggestions/domain/suggestion.dart
import 'package:flutter/foundation.dart';

/// Domain model, mirroring `backend/openapi.yaml`.
///
/// The spec is the source of truth for these names; when the two disagree, this
/// file is the one that is wrong. See `CLAUDE.md`.

/// Where the messages came from.
enum SuggestionSource {
  /// Generated for this request.
  llm,

  /// A previous generation for an equivalent query.
  cache,

  /// Curated messages: the model could not be reached.
  fallback;

  /// Tolerant of a value this build has never heard of.
  ///
  /// A server that learns a new source must not break an app that has not been
  /// updated yet — and the app does not branch on this anyway. It branches on
  /// `degraded`, which is a boolean and cannot acquire new members.
  static SuggestionSource parse(String? raw) => switch (raw) {
    'llm' => SuggestionSource.llm,
    'cache' => SuggestionSource.cache,
    'fallback' => SuggestionSource.fallback,
    _ => SuggestionSource.llm,
  };
}

@immutable
class Suggestion {
  const Suggestion({required this.id, required this.text});

  final String id;
  final String text;

  @override
  bool operator ==(Object other) =>
      other is Suggestion && other.id == id && other.text == text;

  @override
  int get hashCode => Object.hash(id, text);
}

@immutable
class SuggestionResult {
  const SuggestionResult({
    required this.suggestions,
    required this.source,
    required this.degraded,
  });

  final List<Suggestion> suggestions;
  final SuggestionSource source;

  /// True when the LLM could not serve this request.
  ///
  /// The single flag the UI branches on. Kept separate from [source] because it
  /// is a stable boolean, while the set of sources may grow.
  final bool degraded;

  // `meta.model` and `meta.requestId` are deliberately not modelled here. They
  // are server diagnostics: the app has no use for a model name, and the
  // correlation id it would show a user comes from `ServerFailure`, not from a
  // successful response. A client should model what it uses.
}

/// What the user asked for.
@immutable
class SuggestionQuery {
  const SuggestionQuery({
    required this.occasion,
    required this.relationship,
    this.tone = 'warm',
    this.count = 3,
  });

  final String occasion;
  final String relationship;
  final String tone;
  final int count;
}

/// Vocabulary offered by the pickers.
///
/// Fetched from the server so a new occasion does not need a new app build, but
/// the app carries defaults: the picker is a convenience, and failing to load it
/// must not stop someone typing their own occasion.
@immutable
class SuggestionOptions {
  const SuggestionOptions({
    required this.occasions,
    required this.relationships,
    required this.tones,
  });

  static const SuggestionOptions defaults = SuggestionOptions(
    occasions: [
      'Birthday',
      'Wedding',
      'Anniversary',
      'Thank you',
      'Congratulations',
      'Graduation',
      'Holidays',
      'New baby',
      'Get well',
      'Retirement',
      'Housewarming',
      'Farewell',
    ],
    relationships: [
      'Friend',
      'Colleague',
      'Manager',
      'Client',
      'Partner',
      'Mother',
      'Father',
      'Sister',
      'Brother',
      'Daughter',
      'Son',
      'Grandparent',
      'Teacher',
    ],
    tones: ['warm', 'funny', 'formal', 'heartfelt'],
  );

  final List<String> occasions;
  final List<String> relationships;
  final List<String> tones;
}

File: backend/src/domain/suggestion.ts
/** Core domain types. Deliberately free of HTTP, Express and provider concepts. */

/**
 * Where the messages the client received actually came from.
 *
 * This is part of the public contract: the client renders a different affordance
 * for `fallback` (a "showing generic suggestions" notice) than for `llm`.
 */
export const SUGGESTION_SOURCES = ['llm', 'cache', 'fallback'] as const;
export type SuggestionSource = (typeof SUGGESTION_SOURCES)[number];

export interface Suggestion {
  /** Stable within a response; used as a list key by the client. */
  readonly id: string;
  readonly text: string;
}

export const TONES = ['warm', 'funny', 'formal', 'heartfelt'] as const;
export type Tone = (typeof TONES)[number];

export interface SuggestionQuery {
  readonly occasion: string;
  readonly relationship: string;
  readonly tone: Tone;
  readonly count: number;
}

export interface SuggestionResult {
  readonly suggestions: readonly Suggestion[];
  readonly source: SuggestionSource;
  /** True whenever the LLM could not serve this request. */
  readonly degraded: boolean;
  /** Absent on the fallback path — there is no model behind those messages. */
  readonly model: string | undefined;
}

/** The vocabulary a client offers in its pickers. */
export interface SuggestionOptions {
  readonly occasions: readonly string[];
  readonly relationships: readonly string[];
  readonly tones: readonly string[];
}

/** Bounds shared by input validation and output sanitisation. */
export const SUGGESTION_LIMITS = {
  minCount: 2,
  maxCount: 3,
  /** A gift card message that does not fit on a gift card is not a suggestion. */
  maxTextLength: 180,
  minTextLength: 4,
  maxFieldLength: 50,
} as const;
